### ISSUE TITLE
why serde_json::value::to_value(&10i64).is_u64() == true ?

### ISSUE THREAD

**Issue by fengcen** [serde-rs/json#146]
`serde_json::value::to_value(&10i64)` should be `Value::I64` instead of `Value::U64`

**Comment by nabijaczleweli** [comment]
Probably because it doesn't have a sign

**Comment by oli-obk** [comment]
There's no way for serde_json to distinguish between `i64` and `u64`, mainly because json does not have signs. Maybe `Value` should be changed to have a single `Numeric` variant, which inside contains an object that can be converted to various target types.

**Comment by fengcen** [comment]
Thank you for the information. 
@oli-obk Agree. `Value` should only use `F64` variant. Because `f64` can convert to other numeric types without losing precision.

**Comment by oli-obk** [comment]
You will loose precision for large integer values, so thats a nogo. I meant to create an inner ttype that has accessor methods giving options when the conversion was successful

**Comment by dtolnay** [comment]
The confusing behavior comes from [here](https://github.com/serde-rs/json/blob/b4022a299a4664e6bd983471b4519c6353352a23/json/src/value.rs#L605-L609):

```rust
fn serialize_i64(&mut self, value: i64) -> Result<(), Error> {
    if value < 0 {
        self.value = Value::I64(value);
    } else {
        self.value = Value::U64(value as u64);
    }
    Ok(())
}
```

What if we remove the check and always treat i64 as i64? This is a breaking change so I am tentatively scheduling this for 0.9.0.

**Comment by dtolnay** [comment]
I fixed this (sort of) in #188 by clarifying the meaning of `is_u64`. As we know, JSON does not distinguish between signed and unsigned numbers. The new `Value` uses an opaque `Number` type which is a JSON number but is not guaranteed to be any particular Rust number type.

```rust
pub enum Value {
    Null,
    Bool(bool),
    Number(Number),
    String(String),
    Array(Vec<Value>),
    Object(Map<String, Value>),
}
```

The `is_u64` method is commented:

```rust
/// Returns true if the `Value` is a number that can be represented by u64.
```

So `to_value(10i64).is_u64()` is true as before but `to_value(-10i64).is_u64()` is false because -10 cannot be represented as u64. I think this is the right balance of practical to use while still respecting the vaguery of JSON.
